### ISSUE TITLE
Is it possible to run OpenMM on more than node using CPU?

### ISSUE THREAD

**Issue by hima111997** [openmm/openmm#4072]
Hi

I have access to a cluster with 24 cores for each node. I was wondering if it is possible to run a simulation using OpenMM on more than one node. and if it possible, can you tell me what is the correct way to run it?

Thanks

**Comment by peastman** [comment]
The core OpenMM library only supports running a simulation on one node.  There are higher level libraries that will use multiple nodes.  For example, [openmmtools](https://github.com/choderalab/openmmtools) has an implementation of replica exchange that can run each replica on a different node.  And the [metadynamics](http://docs.openmm.org/latest/api-python/generated/openmm.app.metadynamics.Metadynamics.html#openmm.app.metadynamics.Metadynamics) implementation in OpenMM can have walkers running on multiple nodes, as long as they can all access a shared directory for writing biases to.  And of course, you can always just run a separate simulation on each node.  For many cases that's the most efficient thing to do, though of course it depends on your application.

**Comment by hima111997** [comment]
okay thank you. 
i dont have experience in parallelizing softwares, but should not it be faster if a single simulation could run on more cores?   بتاريخ الثلاثاء، 16 أيار 2023 في 08:23:35 م غرينتش+2، Peter Eastman ***@***.***> كتب:  
 
 


The core OpenMM library only supports running a simulation on one node. There are higher level libraries that will use multiple nodes. For example, openmmtools has an implementation of replica exchange that can run each replica on a different node. And the metadynamics implementation in OpenMM can have walkers running on multiple nodes, as long as they can all access a shared directory for writing biases to. And of course, you can always just run a separate simulation on each node. For many cases that's the most efficient thing to do, though of course it depends on your application.

**Comment by peastman** [comment]
It does run on multiple cores.  But only on one node.  Splitting a single simulation across multiple nodes requires different techniques.

You can set the `OPENMM_CPU_THREADS` environment variable to tell it how many threads to create.  By default it creates as many threads as you have cores.  Sometimes it's useful to create fewer threads, since performance doesn't scale quite linearly with the number of threads.  Since you have 24 cores, it might be more efficient to run two simulations at once that each use 12 threads, or three simulations that each use 8.  That lets you stay closer to the linear scaling regime and get more throughput overall.

**Comment by hima111997** [comment]
So if I set OPENMM_CPU_THREADS to 1 and used 2 nodes will it run using the 48 core for single simulation?

Sent from Yahoo Mail on Android 
 
  On Tue, 16 May 2023 at 9:20 pm, Peter ***@***.***> wrote:   


It does run on multiple cores. But only on one node. Splitting a single simulation across multiple nodes requires different techniques.

You can set the OPENMM_CPU_THREADS environment variable to tell it how many threads to create. By default it creates as many threads as you have cores. Sometimes it's useful to create fewer threads, since performance doesn't scale quite linearly with the number of threads. Since you have 24 cores, it might be more efficient to run two simulations at once that each use 12 threads, or three simulations that each use 8. That lets you stay closer to the linear scaling regime and get more throughput overall.

**Comment by sef43** [comment]
> So if I set OPENMM_CPU_THREADS to 1 and used 2 nodes will it run using the 48 core for single simulation? 

No, It will run a single simulation on 1 core on 1 node. The OpenMM CPU platform uses a shared memory parallel programming model. It uses 1 process, which can have multiple threads. When you set `OPENMM_CPU_THREADS` it controls the number of threads to use. 1 process can only run on 1 node, but it can use multiple cores on the same node. It is not possible to run a single OpenMM process on more than a single node. 

Note that other HPC programs use distributed memory parallel program models (usually MPI) which enable them to run on mode that one node. OpenMM does not have this functionality.

See here for an overview of the parallel computing terms I have mentioned: https://hpc-wiki.info/hpc/Parallel_Programming

**Comment by hima111997** [comment]
Okay thank you

Sent from Yahoo Mail on Android 
 
  On Wed, 17 May 2023 at 12:41 pm, Stephen ***@***.***> wrote:   



So if I set OPENMM_CPU_THREADS to 1 and used 2 nodes will it run using the 48 core for single simulation?


No, It will run a single simulation on 1 core on 1 node. The OpenMM CPU platform uses a shared memory parallel programming model. It uses 1 process, which can have multiple threads. When you set OPENMM_CPU_THREADS it controls the number of threads to use. 1 process can only run on 1 node, but it can use multiple cores on the same node. It is not possible to run a single OpenMM process on more than a single node.

Note that other HPC programs use distributed memory parallel program models (usually MPI) which enable them to run on mode that one node. OpenMM does not have this functionality.

See here for an overview of the parallel computing terms I have mentioned: https://hpc-wiki.info/hpc/Parallel_Programming
